### PR TITLE
#16 Interactive mode

### PR DESCRIPTION
**Pull request by outime**

### CODE REVIEW

File: piu/error_handling.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Stores the exception in a temporary file and returns its filename
+    """
+
+    tracebacks = format_exception(etype=type(exception),
+                                  value=exception,
+                                  tb=exception.__traceback__)  # type: [str]
+
+    content = ''.join(tracebacks)
+
+    with NamedTemporaryFile(prefix="senza-traceback-", delete=False) as error_file:
```
Comment (hjacobs) [inline piu/error_handling.py]: senza?

File: piu/cli.py [hunk 2]
```diff
@@ -177,24 +180,58 @@ def cli(ctx, config_file):
...
+@click.option('--clip', help='Copy SSH command into clipboard', is_flag=True, default=False)
+@click.option('--connect', help='Directly connect to the host', envvar='PIU_CONNECT', is_flag=True, default=False)
 @click.pass_obj
-def request_access(obj, host, user, password, even_url, odd_host, reason, reason_cont, insecure, lifetime,
-                   clip, connect):
+def request_access(obj, host, reason, reason_cont, user, password, even_url, odd_host, lifetime, interactive,
+                   insecure, clip, connect):
     '''Request SSH access to a single host'''
 
+    if interactive:
```
Comment (hjacobs) [inline piu/cli.py:203]: can you move this whole part into a function?

**Comment by hjacobs** [comment]
Please add tests.

File: piu/cli.py [hunk 3]
```diff
@@ -177,24 +180,58 @@ def cli(ctx, config_file):
...
+        ec2 = boto3.resource('ec2')
+        reservations = ec2.instances.filter(
```
Comment (hjacobs) [inline piu/cli.py]: you need to add paging support here

Comment (hjacobs) [inline piu/cli.py]: or maybe not (?) --- I usually use the low-level "clients"

File: piu/cli.py [hunk 4]
```diff
@@ -177,24 +180,58 @@ def cli(ctx, config_file):
...
+@click.option('--connect', help='Directly connect to the host', envvar='PIU_CONNECT', is_flag=True, default=False)
 @click.pass_obj
-def request_access(obj, host, user, password, even_url, odd_host, reason, reason_cont, insecure, lifetime,
-                   clip, connect):
+def request_access(obj, host, reason, reason_cont, user, password, even_url, odd_host, lifetime, interactive,
+                   insecure, clip, connect):
     '''Request SSH access to a single host'''
 
+    if interactive:
+        ec2 = boto3.resource('ec2')
```
Comment (hjacobs) [inline piu/cli.py]: assumes the current region `AWS_DEFAULT_REGION` or `~/.aws/config` is correct --- that differs from other STUPS' tools

File: piu/cli.py [hunk 5]
```diff
@@ -177,24 +180,58 @@ def cli(ctx, config_file):
...
+                       Filters=[{'Name': 'instance-state-name', 'Values': ['running']}])
+        stack_name = stack_version = None
+        instance_list = []
+        for r in reservations:
+            tags = r.tags
+            if not tags:
+                continue
+            for d in tags:
+                d_k, d_v = d['Key'], d['Value']
+                if d_k == 'StackName':
```
Comment (hjacobs) [inline piu/cli.py]: We should add the "Name" tag too

**Comment by outime** [comment]
@hjacobs thanks for all the suggestions, I find them reasonable and will be done next. I had tests in mind but wanted to make sure about the implementation first.

**Comment by outime** [comment]
@hjacobs tests pass in my local computer (tried with Python 3.4.x and 3.5.x) but seems stuck in Travis CI. Would you mind taking a look to that as well as the rest of the code? On the other hand, seems that resources handle pagination for you:

> A collection seamlessly handles pagination for you, making it possible to easily iterate over all items from all pages of data. ([source](http://boto3.readthedocs.io/en/latest/guide/collections.html#overview))

File: piu/cli.py [hunk 6]
```diff
@@ -280,8 +280,8 @@ def request_access(obj, host, reason, reason_cont, user, password, even_url, odd
 
 
 def request_access_interactive():
-    region_name = click.prompt('AWS region', default=os.getenv('PIU_REGION') or
-                               subprocess.getoutput('aws configure get region'))
+    region_name = os.getenv('PIU_REGION') or click.prompt('AWS region',
```
Comment (hjacobs) [inline piu/cli.py]: ohoh, please check how Senza does it: `AWS_DEFAULT_REGION` and `~/.aws/config` is honored

**Comment by outime** [comment]
@hjacobs done, is this better?

**Comment by hjacobs** [comment]
@tuxlife any opinion on this?

**Comment by tuxlife** [comment]
@outime Why did you need a interactive mode?

**Comment by outime** [comment]
@tuxlife it's much more comfortable to use Più this way since you don't have to look up for the private IP of the instance with Senza, copy it, then execute Più, etc. Here you get a list of instances for the specified region and choose the one you want easily without caring about IP addresses. I use it with the `--connect` flag which makes it even easier to use as it connects right away.

On the other hand, I've met some Zalando folks who are using a custom wrapper with similar functionality. I also use [one](https://github.com/ramirantala/utils/blob/master/pau). It feels natural to have this feature built-in to have similar workflow.

**Comment by szuecs** [comment]
:+1:

**Comment by tuxlife** [comment]
:+1:

**Comment by outime** [comment]
Will submit PR for docs when the newer package is published :)

**Comment by outime** [comment]
Could any of the maintainers update the PyPI package whenever you have some time? I've also written some docs: https://github.com/zalando-stups/documentation/pull/89

**Comment by hjacobs** [comment]
@outime this just does not work on Travis nor our CD: 

```
tests/test_cli.py::test_interactive_success 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

**Comment by outime** [comment]
@hjacobs I was on holidays so I couldn't check this. Now I see it's fixed (thanks @mikkeloscar) so hopefully the package can now be updated.

**Pull request merged**